File: pkg/rtweekend/aabb.go
// Axis-Aligned bounding boxes
package rtweekend

import (
	"math"
)

type AABB struct {
	minimum Point3
	maximum Point3
}

/* Returns a pointer for AABB (Axis-Aligned Bounding Boxes) object
   Params
       minimum Point3  Coordinate of passing interval
       maximum Point3  Coordinate of passing interval
*/
func NewAABB(minimum, maximum Point3) *AABB {
	aabb := new(AABB)

	aabb.minimum = minimum
	aabb.maximum = maximum

	return aabb
}

func SurroundingBox(box0, box1 *AABB) *AABB {
	small := NewPoint3(
		math.Min(box0.minimum.X(), box1.minimum.X()),
		math.Min(box0.minimum.Y(), box1.minimum.Y()),
		math.Min(box0.minimum.Z(), box1.minimum.Z()),
	)
	big := NewPoint3(
		math.Min(box0.maximum.X(), box1.maximum.X()),
		math.Min(box0.maximum.Y(), box1.maximum.Y()),
		math.Min(box0.maximum.Z(), box1.maximum.Z()),
	)

	return NewAABB(*small, *big)
}

// Returns minimum coordinate of passing interval in AABB
func (aabb *AABB) Min() *vec3.Point3 {
	return &aabb.minimum
}

// Returns maximum coordinate of passing interval in AABB
func (aabb *AABB) Max() *vec3.Point3 {
	return &aabb.maximum
}

func (aabb AABB) Hit(r *ray.Ray, t_min, t_max float64, rec *HitRecord) bool {
	var t0, t1, invD float64

	swap := func(t0, t1 *float64) {
		*t0, *t1 = *t1, *t0
	}

	for a := 0; a < 3; a++ {
		invD = 1.0 / r.Direction().E(a)
		t0 = (aabb.minimum.E(a) - r.Origin().E(a)) * invD
		t1 = (aabb.maximum.E(a) - r.Origin().E(a)) * invD

		if invD < 1 {
			swap(&t0, &t1)
		}

		if t0 > t_min {
			t_min = t0
		}

		if t_max > t1 {
			t_max = t1
		}

		if t_max <= t_min {
			return false
		}
	}

	return true
}

func (aabb AABB) BoundingBox(time0, time1 float64, outputBox *AABB) bool {
	return false
}

File: src/lib/aabb.rs
use crate::{max, min};
use crate::{Point3, Ray};
use std::mem;

// Axis-Aligned Bounding Boxes
#[derive(Clone, Copy, PartialEq)]
pub struct Aabb {
    minimum: Point3,
    maximum: Point3,
}

impl Aabb {
    #[inline]
    pub fn new(minimum: Point3, maximum: Point3) -> Self {
        Self { minimum, maximum }
    }

    #[inline]
    pub fn min(&self) -> Point3 {
        self.minimum
    }

    #[inline]
    pub fn max(&self) -> Point3 {
        self.maximum
    }

    #[inline]
    pub fn hit(
        &self,
        r: &Ray,
        mut t_min: f64,
        mut t_max: f64,
    ) -> Option<(Option<f64>, Option<f64>)> {
        for a in 0..3 {
            let inv_d = 1. / r.direction()[a];
            let mut t0 = min(
                (self.min()[a] - r.origin()[a]) * inv_d,
                (self.max()[a] - r.origin()[a]) * inv_d,
            );
            let mut t1 = max(
                (self.minimum[a] - r.origin()[a]) / r.direction()[a],
                (self.maximum[a] - r.origin()[a]) / r.direction()[a],
            );

            if inv_d < 0. {
                mem::swap(&mut t0, &mut t1);
            }

            t_min = max(t0, t_min);
            t_max = min(t1, t_max);

            if t_max <= t_min {
                return Some((None, None));
            }
        }

        Some((Some(t_min), Some(t_max)))
    }
}

pub fn surrounding_box(box0: Aabb, box1: Aabb) -> Aabb {
    let small = Point3::new(
        min(box0.min().x(), box1.min().x()),
        min(box0.min().y(), box1.min().y()),
        min(box0.min().z(), box1.min().z()),
    );
    let big = Point3::new(
        max(box0.min().x(), box1.min().x()),
        max(box0.min().y(), box1.min().y()),
        max(box0.min().z(), box1.min().z()),
    );

    Aabb::new(small, big)
}
